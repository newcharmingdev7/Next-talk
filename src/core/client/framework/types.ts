// TODO: (@cvle) Extract useful common types into its own package.
export { Omit, Overwrite, PropTypesOf } from "coral-ui/types";
export { DeepPartial } from "coral-common/types";

export type RelayEnumLiteral<T> = keyof T | "%future added value";
