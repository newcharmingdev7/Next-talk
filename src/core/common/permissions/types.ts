export type UserRole =
  | "%future added value"
  | "COMMENTER"
  | "MEMBER"
  | "STAFF"
  | "MODERATOR"
  | "ADMIN";

export interface User {
  id: string;
  role: UserRole;
  moderationScopes?: null | {
    scoped?: boolean;
    siteIDs?: string[];
  };
}

export type Moderator = User & {
  role: "MODERATOR";
};

export type LTEModerator = User & {
  role: "MODERATOR" | "STAFF" | "MEMBER" | "COMMENTER";
};

export const isModerator = (user: User): user is Moderator => {
  return user.role === "MODERATOR";
};

export const isLTEModerator = (user: User): user is LTEModerator => {
  return user.role !== "ADMIN";
};
