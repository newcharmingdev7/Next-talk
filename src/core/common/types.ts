export type Diff<T extends keyof any, U extends keyof any> = ({ [P in T]: P } &
  { [P in U]: never } & { [x: string]: never })[T];

export type Omit<U, K extends keyof U> = Pick<U, Diff<keyof U, K>>;

export type Overwrite<T, U> = Pick<T, Diff<keyof T, keyof U>> & U;

export type Sub<T, U> = Pick<T, Diff<keyof T, keyof U>>;

/**
 * Make all properties in T writeable
 */
export type Writeable<T> = { -readonly [P in keyof T]: T[P] };

export type Promiseable<T> = Promise<T> | T;
