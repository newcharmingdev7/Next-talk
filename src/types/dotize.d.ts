declare module "dotize" {
  export function convert(obj: Object): Record<string, any>
}
