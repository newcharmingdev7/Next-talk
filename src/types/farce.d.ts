declare module "farce" {
  export class BrowserProtocol {}
  export class queryMiddleware {}
}
