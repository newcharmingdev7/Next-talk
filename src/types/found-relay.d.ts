declare module "found-relay" {
  import { Environment } from "relay-runtime";
  export class Resolver {
    constructor(relayEnvironment: Environment);
  }
}
