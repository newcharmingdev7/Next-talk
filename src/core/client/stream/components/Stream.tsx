import * as React from "react";
import { StatelessComponent } from "react";

import Logo from "talk-stream/components/Logo";
import CommentContainer from "talk-stream/containers/CommentContainer";
import PostCommentFormContainer from "talk-stream/containers/PostCommentFormContainer";
import { Button } from "talk-ui/components";

export interface StreamProps {
  id: string;
  isClosed: boolean;
  comments: ReadonlyArray<{ id: string }> | null;
  onLoadMore: () => void;
  hasMore: boolean;
}

const Stream: StatelessComponent<StreamProps> = props => {
  if (props.comments === null) {
    // TODO: (@cvle) What's the reason for this being null?
    return <div>Comments unavailable</div>;
  }
  return (
    <div>
      <Logo gutterBottom />
      <PostCommentFormContainer assetID={props.id} />
      {props.comments.map(comment => (
        <CommentContainer key={comment.id} data={comment} gutterBottom />
      ))}
      {props.hasMore && (
        <Button
          className={"talk-stream--loadmore"}
          onClick={props.onLoadMore}
          secondary
          invert
          fullWidth
        >
          Load More
        </Button>
      )}
    </div>
  );
};

export default Stream;
