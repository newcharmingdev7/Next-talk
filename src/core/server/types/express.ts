import { Request } from "express";

import { Tenant } from "talk-server/models/tenant";
import { User } from "talk-server/models/user";
import TenantCache from "talk-server/services/tenant/cache";

export interface Request extends Request {
  user?: User;
  tenant?: Tenant;
  tenantCache: TenantCache;
}
