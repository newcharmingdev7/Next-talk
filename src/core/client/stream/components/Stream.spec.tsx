import { shallow } from "enzyme";
import { noop } from "lodash";
import React from "react";
import sinon, { SinonSpy } from "sinon";

import { PropTypesOf } from "talk-framework/types";

import Stream from "./Stream";

it("renders correctly", () => {
  const props: PropTypesOf<typeof Stream> = {
    assetID: "asset-id",
    isClosed: false,
    comments: [{ id: "comment-1" }, { id: "comment-2" }],
    onLoadMore: noop,
    disableLoadMore: false,
    hasMore: false,
    user: null,
  };
  const wrapper = shallow(<Stream {...props} />);
  expect(wrapper).toMatchSnapshot();
});

describe("when use is logged in", () => {
  it("renders correctly", () => {
    const props: PropTypesOf<typeof Stream> = {
      assetID: "asset-id",
      isClosed: false,
      comments: [{ id: "comment-1" }, { id: "comment-2" }],
      onLoadMore: noop,
      disableLoadMore: false,
      hasMore: false,
      user: {},
    };
    const wrapper = shallow(<Stream {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("when there is more", () => {
  const props: PropTypesOf<typeof Stream> = {
    assetID: "asset-id",
    isClosed: false,
    comments: [{ id: "comment-1" }, { id: "comment-2" }],
    onLoadMore: sinon.spy(),
    disableLoadMore: false,
    hasMore: true,
    user: null,
  };

  const wrapper = shallow(<Stream {...props} />);
  it("renders a load more button", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("calls onLoadMore", () => {
    wrapper.find("#talk-comments-stream-loadMore").simulate("click");
    expect((props.onLoadMore as SinonSpy).calledOnce).toBe(true);
  });

  const wrapperDisabledButton = shallow(<Stream {...props} disableLoadMore />);
  it("disables load more button", () => {
    expect(wrapperDisabledButton).toMatchSnapshot();
  });
});
