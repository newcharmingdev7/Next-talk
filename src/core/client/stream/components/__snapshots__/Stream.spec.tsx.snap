// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Stream-root"
>
  <Logo
    gutterBottom={true}
  />
  <withContext(createMutationContainer(PostCommentFormContainer))
    assetID="asset-id"
  />
  <Flex
    aria-live="polite"
    direction="column"
    id="talk-comments-stream-log"
    itemGutter={true}
    role="log"
  >
    <Flex
      direction="column"
      itemGutter={true}
      key="comment-1"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-1",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-1",
          }
        }
      />
    </Flex>
    <Flex
      direction="column"
      itemGutter={true}
      key="comment-2"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-2",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-2",
          }
        }
      />
    </Flex>
  </Flex>
</div>
`;

exports[`when there is more disables load more button 1`] = `
<div
  className="Stream-root"
>
  <Logo
    gutterBottom={true}
  />
  <withContext(createMutationContainer(PostCommentFormContainer))
    assetID="asset-id"
  />
  <Flex
    aria-live="polite"
    direction="column"
    id="talk-comments-stream-log"
    itemGutter={true}
    role="log"
  >
    <Flex
      direction="column"
      itemGutter={true}
      key="comment-1"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-1",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-1",
          }
        }
      />
    </Flex>
    <Flex
      direction="column"
      itemGutter={true}
      key="comment-2"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-2",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-2",
          }
        }
      />
    </Flex>
    <Localized
      id="comments-stream-loadMore"
    >
      <withPropsOnChange(Button)
        aria-controls="talk-comments-stream-log"
        disabled={true}
        fullWidth={true}
        id="talk-comments-stream-loadMore"
        invert={true}
        onClick={[Function]}
        secondary={true}
      >
        Load More
      </withPropsOnChange(Button)>
    </Localized>
  </Flex>
</div>
`;

exports[`when there is more renders a load more button 1`] = `
<div
  className="Stream-root"
>
  <Logo
    gutterBottom={true}
  />
  <withContext(createMutationContainer(PostCommentFormContainer))
    assetID="asset-id"
  />
  <Flex
    aria-live="polite"
    direction="column"
    id="talk-comments-stream-log"
    itemGutter={true}
    role="log"
  >
    <Flex
      direction="column"
      itemGutter={true}
      key="comment-1"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-1",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-1",
          }
        }
      />
    </Flex>
    <Flex
      direction="column"
      itemGutter={true}
      key="comment-2"
    >
      <Relay(CommentContainer)
        data={
          Object {
            "id": "comment-2",
          }
        }
      />
      <Relay(ReplyListContainer)
        comment={
          Object {
            "id": "comment-2",
          }
        }
      />
    </Flex>
    <Localized
      id="comments-stream-loadMore"
    >
      <withPropsOnChange(Button)
        aria-controls="talk-comments-stream-log"
        disabled={false}
        fullWidth={true}
        id="talk-comments-stream-loadMore"
        invert={true}
        onClick={[Function]}
        secondary={true}
      >
        Load More
      </withPropsOnChange(Button)>
    </Localized>
  </Flex>
</div>
`;
